javascript/modern-js.md

```diff
@@ -190,69 +206,153 @@ showResult(name, subject, grade);
-    return parsedData;
-    } catch (err) {
-      // ...the code in this block will execute. The error that has been created will now be inserted into `err`
-      console.log('Oops, something went wrong!', err);
+async function fetchData(url) {
+  try {
+    const response = await fetch(url);
+    if (response.ok) {
+      return await response.json();
     }
+    throw new Error(`HTTP Error ${response.status}:  ${response.statusText}`);
+  } catch (error) {
+    console.error(`Oops, something went wrong: ${error.message}`);
   }
+}
 ```
 
+If an error occurs inside the `fetchData()` function the `catch` block executes, which logs the error to the console. With the error now "handled" (i.e. "consumed") by the `catch` block, the function subsequently returns with `undefined` as its return value. The calling function, in our example `fetchAndRender()`, no longer sees an error, as `fetchData()` already "handled" it. Instead, it now receives the value `undefined` and assigns that to the `data` variable. Obviously this will result in undefined behaviour.
+
+Here is how it is worded in [mdn web docs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/async_function#description):
+
+> _Each time when an async function is called, it returns a new Promise which will be resolved with the value returned by the async function, or rejected with an exception uncaught within the async function._
+
+In the example function above the exception _is_ caught and therefore it returns a resolved promise with the value of `undefined` (because `fetchData()` has no explicit `return` value) rather than a rejected promise with the error.
+
+If you would like your function to log an error to the console while still passing on the error to the calling function you can "rethrow" the error inside the `catch` block, like this:
+
+```js
+async function fetchData(url) {
+  try {
+    const response = await fetch(url);
+    if (response.ok) {
+      return await response.json();
+    }
+    throw new Error(`HTTP Error ${response.status}:  ${response.statusText}`);
+  } catch (error) {
+    console.error(`Oops, something went wrong: ${error.message}`);
+    throw error; // <== rethrow the error
+  }
+}
+```
+
+Now the added `throw` in `fetchData()` is uncaught in the function itself resulting in a rejected promise that can be caught in the calling function.
+
+Finally, you may have noticed that we added an `async` keyword when returning the result from `response.json()`. The `response.json()` method may return a rejected promise in case the response text cannot be parsed as valid JSON. The added the `await` keyword ensures that an [exception is thrown](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/await#exceptions) inside the `try` block of `fetchData()` if `response.json()` returns a rejected promise. This will then be caught by the `catch` block which causes parse error to be logged to the console before it is rethrown.
```

Comment:
More correct?

```suggestion
Finally, you may have noticed that we added an `await` keyword when returning the result from `response.json()`. The `response.json()` method may return a rejected promise in case the response text cannot be parsed as valid JSON. The added `await` keyword ensures that an [exception is thrown](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/await#exceptions) inside the `try` block of `fetchData()` if `response.json()` returns a rejected promise. This will then be caught by the `catch` block which causes parse error to be logged to the console before it is rethrown.
```